Bluetooth: Controller: Replace BT_CTLR with HAS_BT_CTLR

jhedberg opened the pull request:
Deprecate BT_CTLR, and add a new HAS_BT_CTLR as a virtual option which specific users (like BT_LL_SW_SPLIT) select. This also means that we can remove all places that were forcefully enabling the BT_CTLR option, and instead we now depend on devicetree to get some local LL HCI driver enabled which in turn selects the HAS_BT_CTLR option. Also, any place that was explicitly disabling BT_CTLR now needs to be replaced with the only previous upstream user of BT_CTLR getting disabled, namely BT_LL_SW_SPLIT.

jhedberg committed:
Bluetooth: Controller: Replace BT_CTLR with HAS_BT_CTLR

Deprecate BT_CTLR, and add a new HAS_BT_CTLR as a virtual option which
specific users (like BT_LL_SW_SPLIT) select. This also means that we can
remove all places that were forcefully enabling the BT_CTLR option, and
instead we now depend on devicetree to get some local LL HCI driver
enabled which in turn also enables the HAS_BT_CTLR option.

Signed-off-by: Johan Hedberg <johan.hedberg@silabs.com>

cvinayak commented on boards/shields/mikroe_ble_tiny_click/Kconfig.defconfig:
```diff
@@ -3,4 +3,4 @@
 CONFIG_BT_HCI=y
 CONFIG_BT_HCI_ACL_FLOW_CONTROL=n
-CONFIG_BT_CTLR=n
+CONFIG_BT_LL_SW_SPLIT=n
```
This is odd. We should not need to disable? Does this mean the shield supports Controller at SoC, but not on boards/sheilds?

cvinayak commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -121,38 +121,33 @@ config BT_CTLR_SUBRATING_SUPPORT
-
-config BT_CTLR_HCI
-	bool "Host Controller Interface (HCI)"
+config BT_LL_SW_SPLIT_HCI
+	bool "Host Controller Interface (HCI) for the split link layer"
```
```suggestion
config BT_CTLR_HCI
	bool "Host Controller Interface (HCI), Controller-side"
```

cvinayak commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -121,38 +121,33 @@ config BT_CTLR_SUBRATING_SUPPORT
+	bool "Host Controller Interface (HCI) for the split link layer"
 	default y
-	depends on DT_HAS_ZEPHYR_BT_HCI_LL_SW_SPLIT_ENABLED
+	depends on BT_LL_SW_SPLIT
```
This is open to discussion on how to include downstream Controllers/Blobs (with/without HCI layer)

cvinayak reviewed (dismissed):
Approved `BT_CTLR` changes.
Move out removal (discussion) of BT_CTLR_HCI to a separate PR.

jhedberg commented on subsys/bluetooth/controller/CMakeLists.txt:
```diff
@@ -29,7 +29,7 @@ zephyr_library_sources(
 
 zephyr_library_sources_ifdef(
-  CONFIG_BT_CTLR_HCI
+  CONFIG_BT_LL_SW_SPLIT_HCI
```
> Suggest to retain `BT_CTLR_HCI` and move the block to Controller specific cmake list file.

What file are you referring to? E.g. the `hci` or `ll_sw` subdirectories don't have their own `CMakeLists.txt` files.

What should the Kconfig dependencies of the option be? Right now it's LL_SW, which is wrong based on your description of what this option should be.

cvinayak replied:
> What file are you referring to? E.g. the hci or ll_sw subdirectories don't have their own CMakeLists.txt files.

Since there is only one Controller implementation currently in the `subsys/bluetooth/controller` this is in this CMakelists.txt. I do not see these files ever reused by other Controllers (say, blobs with functional Link Layer interface). But a new CMakelists.txt file can be added inside `hci` folder have:

```
add_subdirectory_ifdef(
  CONFIG_BT_CTLR_HCI
  hci
)
```

Vendor specific HCI implementations under `ll_sw/openisa/hci`  etc.. should also be moved from this parent CMakelists.txt into `hci` folder CMakelists.txt.

>What should the Kconfig dependencies of the option be? 

```
    depends on BT_CTLR
```

jhedberg replied:
That said, I did fix the dependency like you suggested.

jhedberg committed:
doc: migration-guide-4.1: Add mention of deprecated BT_CTLR

Mention that BT_CTLR has been deprecated, and explain the new option that
replaces it.

Signed-off-by: Johan Hedberg <johan.hedberg@silabs.com>

rugeGerritsen commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -121,38 +121,39 @@ config BT_CTLR_SUBRATING_SUPPORT
-if BT_CTLR
+# Virtual option that all local LL implemenations should select
+config HAS_BT_CTLR
+	bool
```
Is it correct to say, that we want to let HAS_BT_CTLR have the default value
```
DT_CHOSEN_ZEPHYR_HCI := zephyr,zephyr,bt-hci

config HAS_BT_CTLR
	bool
	default $(dt_chosen_enabled,$(DT_CHOSEN_ZEPHYR_HCI))
```
If that is the case, then I would assume that we don't need `select HAS_BT_CTLR` in any place. We could even consider getting rid of `config HAS_BT_CTLR` completely.

jhedberg replied:
The DT chosen is used for *all* HCI drivers, whereas `BT_CTLR`/`HAS_BT_CTLR` is specifically for situations where you have a local link layer, so the above doesn't really work.

rugeGerritsen replied:
Thanks, you are right

rugeGerritsen commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -121,38 +121,39 @@ config BT_CTLR_SUBRATING_SUPPORT
 
 config BT_LL_SW_SPLIT
 	bool "Software-based BLE Link Layer"
+	default y
 	depends on DT_HAS_ZEPHYR_BT_HCI_LL_SW_SPLIT_ENABLED
```
To make integration simpler with the SoftDevice Controller, I believe we should change this to:
```suggestion
	default y if DT_HAS_ZEPHYR_BT_HCI_LL_SW_SPLIT_ENABLED
```

jhedberg replied:
That doesn't seem right to me. Not having the explicit dependency would mean that you could enable the option even though the corresponding DT node isn't enabled, which isn't correct. The following pattern is what almost all drivers in Zephyr use in their Kconfig, so I think it makes sense to follow the same practice:
```
default y
depends on DT_HAS_<compat>_ENABLED
```

rugeGerritsen reviewed (commented):
Is it correct to state that this commit header could have been "Bluetooth: Controller: Enable through DTS instead of Kconfig"?

jhedberg commented:
> Is it correct to state that this commit header could have been "Bluetooth: Controller: Enable through DTS instead of Kconfig"?

Not quite. The (upstream) controller's Kconfig option was already depending on DTS, however it also depended on `BT_CTLR` being separately enabled:

https://github.com/zephyrproject-rtos/zephyr/blob/8faa486430db3775ebbb39df76e874a35f742e1d/subsys/bluetooth/controller/Kconfig#L139-L141

rugeGerritsen commented:
> c

I see. But now all controllers/drivers are enabled if they are enabled in DTS?

jhedberg commented:
> I see. But now all controllers/drivers are enabled if they are enabled in DTS?

Yes, this is how driver enabling usually works in Zephyr. You could create trouble if you try to enable two mutually conflicting controllers (e.g. have both a SDC and a LL_SW node in DTS), but that's a bug in your DTS then.

jhedberg commented:
@rugeGerritsen regarding what changes this requires in NCS, you'll need to add `select HAS_BT_CTLR` to `config BT_LL_SOFTDEVICE`, and remove `choice BT_LL_CHOICE` (since that choice doesn't exist anymore). Other than that you should be fine, I think.

jhedberg committed:
doc: bluetooth: Update references to Kconfig options and DTS

Update the instructions for the three possible Bluetooth setups
(controller-only, combined build, host-only) to be in line with how the
features are actually enabled (e.g. device tree for the controller and HCI
drivers).

Signed-off-by: Johan Hedberg <johan.hedberg@silabs.com>

jhedberg commented:
I realized that we can do better than to force each LL/HCI driver to `select HAS_BT_CTLR`. We can actually derive this need from device tree, namely if the driver's HCI binding says `bt-hci-bus: virtual` then we know it's a local controller. Zephyr has Kconfig functions to accomplish this:
```
diff --git a/subsys/bluetooth/controller/Kconfig b/subsys/bluetooth/controller/Kconfig
index fe44e29a75c..18f234807a9 100644
--- a/subsys/bluetooth/controller/Kconfig
+++ b/subsys/bluetooth/controller/Kconfig
@@ -121,9 +121,13 @@ config BT_CTLR_SUBRATING_SUPPORT
 config BT_CTLR_CHANNEL_SOUNDING_SUPPORT
 	bool
 
-# Virtual option that all local LL implemenations should select
+DT_CHOSEN_BT_HCI := zephyr,bt-hci
+DT_BT_HCI_NODE := $(dt_chosen_path,$(DT_CHOSEN_BT_HCI))
+DT_ZEPHYR_BT_HCI_VIRTUAL := $(dt_node_str_prop_equals,$(DT_BT_HCI_NODE),bt-hci-bus,virtual)
+
 config HAS_BT_CTLR
 	bool
+	default y if "$(DT_ZEPHYR_BT_HCI_VIRTUAL)"
 
 config BT_CTLR
 	bool "Bluetooth Controller [DEPRECATED]"
@@ -139,7 +143,6 @@ config BT_LL_SW_SPLIT
 	bool "Software-based BLE Link Layer"
 	default y
 	depends on DT_HAS_ZEPHYR_BT_HCI_LL_SW_SPLIT_ENABLED
-	select HAS_BT_CTLR
 	help
 	  Use Zephyr software BLE Link Layer ULL LLL split implementation.
 ```

jhedberg commented:
There's actually a small challenge with my previous proposal of deriving `HAS_BT_CTLR` from device tree. It means that every driver with "virtual" as the bus must immediately start selecting the appropriate Kconfig feature options. Otherwise there will be host-side features that cannot be enabled due to dependencies like this:
```
subsys/bluetooth/Kconfig:       default y if !HAS_BT_CTLR && !BT_STM32_IPM && !BT_ESP32  && !BT_STM32WBA
subsys/bluetooth/Kconfig:       depends on !HAS_BT_CTLR || BT_CTLR_PHY_UPDATE_SUPPORT
subsys/bluetooth/Kconfig:       depends on !HAS_BT_CTLR || BT_CTLR_DATA_LEN_UPDATE_SUPPORT
subsys/bluetooth/Kconfig:       depends on BT_PER_ADV_SYNC && BT_CONN && (!HAS_BT_CTLR || BT_CTLR_SYNC_TRANSFER_RECEIVER_SUPPORT)
subsys/bluetooth/Kconfig:       depends on (BT_PER_ADV_SYNC || BT_PER_ADV) && BT_CONN && (!HAS_BT_CTLR || BT_CTLR_SYNC_TRANSFER_SENDER_SUPPORT)
subsys/bluetooth/Kconfig:       depends on !HAS_BT_CTLR || BT_CTLR_SCA_UPDATE_SUPPORT
subsys/bluetooth/Kconfig:       depends on !HAS_BT_CTLR || BT_CTLR_LE_POWER_CONTROL_SUPPORT
subsys/bluetooth/Kconfig:       depends on !HAS_BT_CTLR || BT_CTLR_LE_PATH_LOSS_MONITORING_SUPPORT
subsys/bluetooth/Kconfig:       depends on !HAS_BT_CTLR || BT_CTLR_SUBRATING_SUPPORT
subsys/bluetooth/Kconfig:       depends on !HAS_BT_CTLR || BT_CTLR_CHANNEL_SOUNDING_SUPPORT
subsys/bluetooth/Kconfig.adv:   depends on !HAS_BT_CTLR || BT_CTLR_ADV_EXT_SUPPORT
subsys/bluetooth/Kconfig.iso:   depends on !HAS_BT_CTLR || BT_CTLR_PERIPHERAL_ISO_SUPPORT
subsys/bluetooth/Kconfig.iso:   depends on !HAS_BT_CTLR || BT_CTLR_CENTRAL_ISO_SUPPORT
subsys/bluetooth/Kconfig.iso:   depends on !HAS_BT_CTLR || BT_CTLR_ADV_ISO_SUPPORT
subsys/bluetooth/Kconfig.iso:   depends on !HAS_BT_CTLR || BT_CTLR_SYNC_ISO_SUPPORT
subsys/bluetooth/host/Kconfig:  depends on !HAS_BT_CTLR || BT_CTLR_DF_SUPPORT
subsys/bluetooth/host/Kconfig:  depends on !HAS_BT_CTLR || BT_CTLR_DF_CTE_RX_SUPPORT
subsys/bluetooth/host/Kconfig:  depends on !HAS_BT_CTLR || BT_CTLR_DF_CTE_TX_SUPPORT
subsys/bluetooth/host/Kconfig:  depends on !HAS_BT_CTLR || BT_CTLR_DF_CTE_RX_SUPPORT
subsys/bluetooth/host/Kconfig:  depends on !HAS_BT_CTLR || BT_CTLR_DF_CTE_TX_SUPPORT
subsys/bluetooth/host/Kconfig:  default y if HAS_BT_CTLR && !BT_CTLR_ECDH
subsys/bluetooth/mesh/Kconfig:  depends on BT_CTLR_ADV_EXT || !HAS_BT_CTLR
```

@cvinayak btw, all of those `BT_CTLR_*_SUPPORT` options, following the recommended Zephyr Kconfig naming scheme, should start with `HAS_BT_CTLR_*`.

jhedberg commented:
Even better would be if the above feature options were device tree properties instead, or at least derived their default values from device tree properties.

jhedberg commented:
@cvinayak could you take another look and see if you're ok with the changes this now contains? I intentionally left out many of the other things we've discussed, since I see those as independent from this initial change (which unblocks the path to extending other local LLs to use the `BT_CTLR_*` options). I can continue looking at the other enhancements in follow-up PRs.

rugeGerritsen commented on snippets/bt-ll-sw-split/bt-ll-sw-split.conf:
@cvinayak , I guess this conf file is no longer needed? Can be addressed in a separate PR

cvinayak replied:
It enables `CONFIG_BT=y` for now, and makes sense to keep it.

rugeGerritsen reviewed (approved):
Thanks @jhedberg

Thalley commented on snippets/bt-ll-sw-split/README.rst:
Should this file be removed now?

jhedberg replied:
Same thing was asked (and answered here): https://github.com/zephyrproject-rtos/zephyr/pull/82157#discussion_r1865839359

Thalley commented on snippets/bt-ll-sw-split/bt-ll-sw-split.conf:
Should this file be removed now? It's just enabling BT now
